pynest/nest/tests/test_stdp_synapse.py

```diff
@@ -0,0 +1,360 @@
+            if t_next != t_next_:
+                t_next = t_next_
+                handle_pre_spike = False
+                handle_post_spike = False'''
+
+            h = t_next - t
+            Kpre *= exp(-h / self.tau_pre)
+            t = t_next
+
+            if handle_post_spike:
+                # Kpost += 1.    <-- not necessary, will call Kpost_at_time(t) later to compute Kpost for any value t
+                weight = facilitate(weight, Kpre)
+
+            if handle_pre_spike:
+                Kpre += 1.
+                _Kpost = Kpost_at_time(t - self.dendritic_delay, post_spikes, init=self.init_weight, inclusive=False)
+                weight = depress(weight, _Kpost)
+
+            # logging
+            t_log.append(t)
+            w_log.append(weight)
+            Kpre_log.append(Kpre)
+
+        Kpost_log = [Kpost_at_time(t - self.dendritic_delay, post_spikes, init=self.init_weight) for t in t_log]
+        if DEBUG_PLOTS:
+            self.plot_weight_evolution(pre_spikes, post_spikes, t_log, w_log, Kpre_log, Kpost_log,
+                                       fname_snip=fname_snip + "_ref", title_snip="Reference")
+
+        return t_log, w_log
+
+    def plot_weight_evolution(self, pre_spikes, post_spikes, t_log, w_log, Kpre_log=None, Kpost_log=None,
+                              fname_snip="", title_snip=""):
+        fig, ax = plt.subplots(nrows=3)
+
+        n_spikes = len(pre_spikes)
+        for i in range(n_spikes):
+            ax[0].plot(2 * [pre_spikes[i]], [0, 1], linewidth=2, color="blue", alpha=.4)
+        ax[0].set_ylabel("Pre spikes")
+        ax0_ = ax[0].twinx()
+        if Kpre_log:
+            ax0_.plot(t_log, Kpre_log)
+
+        n_spikes = len(post_spikes)
+        for i in range(n_spikes):
+            ax[1].plot(2 * [post_spikes[i]], [0, 1], linewidth=2, color="red", alpha=.4)
+        ax1_ = ax[1].twinx()
+        ax[1].set_ylabel("Post spikes")
+        if Kpost_log:
+            ax1_.plot(t_log, Kpost_log)
+
+        ax[2].plot(t_log, w_log, marker="o", label="nestml")
+        ax[2].set_ylabel("w")
+
+        ax[2].set_xlabel("Time [ms]")
+        for _ax in ax:
+            _ax.grid(which="major", axis="both")
+            _ax.grid(which="minor", axis="x", linestyle=":", alpha=.4)
+            _ax.minorticks_on()
+            _ax.set_xlim(0., self.simulation_duration)
+
+        fig.suptitle(title_snip)
+        fig.savefig("/tmp/nest_stdp_synapse_test" + fname_snip + ".png", dpi=300)
+        plt.close(fig)
+
+    def test_stdp_synapse(self):
+        self.dendritic_delay = float('nan')
+        self.init_params()
+        for self.dendritic_delay in [1., self.resolution]:
+            self.init_params()
+            for self.nest_neuron_model in ["iaf_psc_exp",
+                                           "iaf_cond_exp",
+                                           "iaf_psc_exp_ps"]:
+                fname_snip = "_[nest_neuron_mdl=" + self.nest_neuron_model + "]"
```

Comment:
What is this?

Reply:
A snippet inserted into the filename when saving debug plots to disk.